commons/src/main/java/org/infinispan/commons/api/functional/EntryView.java

```diff
@@ -0,0 +1,122 @@
+package org.infinispan.commons.api.functional;
+
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Entry views expose cached entry information to the user. Depending on the
+ * type of entry view, different operations are available. Currently, three
+ * type of entry views are supported:
+ *
+ * <ul>
+ *    <il>{@link ReadEntryView}: read-only entry view</il>
+ *    <il>{@link WriteEntryView}: write-only entry view</il>
+ *    <il>{@link ReadWriteEntryView}: read-write entry view</il>
+ * </ul>
+ *
+ * @since 8.0
+ */
+public final class EntryView {
+
+   private EntryView() {
+      // Cannot be instantiated, it's just a holder class
+   }
+
+   /**
+    * Expose read-only information about a cache entry potentially associated
+    * with a key in the functional map. Typically, if the key is associated
+    * with a cache entry, that information will include value and optional
+    * {@link MetaParam} information.
+    *
+    * DESIGN RATIONALES:
+    * <ul>
+    *    <li>Why does ReadEntryView expose both get() and find() methods for
+    *    retrieving the value? Convenience. If the caller knows for sure
+    *    that the value will be present, get() offers the convenience of
+    *    retrieving the value directly without having to get an {@link Optional}
+    *    first.
+    *    </li>
+    *    <li>Why have find() return {@link Optional}? Why not get rid of it and only have
+    *    get() method return null? Because nulls are evil. If a value might not
+    *    be there, the user can use {@link Optional} to find out if the value might
+    *    be there and deal with non-present values in a more functional way.
```

Comment:
Just because old fashioned APIs rely on Nulls, we should no expose them directly via our API. The way to represent optional things is `Optional<T>` and not `null`. Nulls are evil.